### DISCUSSION TITLE
certificate based acl

### DISCUSSION THREAD

**Discussion by aze2201** [emqx/emqx#10576] [General]
How to built acl in EMQX enterprise to allow each user to the topic only what mentioned in own certificate common name.

Like mosquitto mqtt
`pattern readwrite %u/#`
or 
`pattern readwrite %c/#`

Looks it is acl example all can compare hard coded string

`{allow, {user, "dashboard"}, subscribe, ["$SYS/#"]}.`

**Comment by SergeTupchiy** [comment]
Hi @aze2201,

We support username/clientid placeholders in topic name, please see some examples below.

EMQX 5:
- client id: `{allow, all, all, ["${clientid}/#"]}.`
- username: `{allow, all, all, ["${username}/#"]}.`

EMQX 4:
- client id:  `{allow, all, pubsub, ["%c/#"]}.`
- username: `{allow, all, pubsub, ["%u/#"]}.`

However, certificate common name is not supported in ACL file for now. We will consider adding this functionality. 

On a side note, certificate common name is currently supported in other authorization back-ends, for example:
- EMQX 5: https://docs.emqx.com/en/enterprise/v5.0/access-control/authz/authz.html#placeholders-in-data-queries
- EMQX 4: https://docs.emqx.com/en/enterprise/v4.4/modules/pgsql_authentication.html#access-control-list

**Comment by aze2201** [comment]
I set up my ```common name: bridge1```
```peer_cert_as_username = cn```
```peer_cert_as_clientid = cn```

So, username should be cn. but ```{allow,  {user, "%u"}, pubsub, ["%u/#"]}.``` should allow me to username is CN allow to pubsub to same username (CN). 

But it does not work.

**Reply by aze2201** [reply]
Means as ```username=CN=bridge1``` can connect or make a bridge to EMQX by named ```topic=bridge```

**Reply by SergeTupchiy** [reply]
This should work, 
the rule should mention `all` users, since `{user, "%u"}` is interpreted literally. 
The complete example for EMQX 4.4:

acl file:
```
{allow, {user, "dashboard"}, subscribe, ["$SYS/#"]}.

{deny, all, subscribe, ["$SYS/#", {eq, "#"}]}.

{allow, all, pubsub, ["%u/#"]}.

{deny, all}.
```
listeners conf:
```
listener.ssl.external.verify = verify_peer
listener.ssl.external.peer_cert_as_username = cn
listener.ssl.external.peer_cert_as_clientid = cn
```

test:
```
$ openssl x509 -noout -subject -in client-cert.pem 
subject=C = CN, ST = hangzhou, O = EMQ, CN = Client

$ emqtt pub  -p 8883 -t 'Client/data/test' -u test -C test --payload '{"val": "auth"}' -r --enable-ssl --cert ./client-cert.pem --key ./client-key.pem --CAfile cacert.pem 
Client test sent CONNECT
Client test sent PUBLISH (Q0, R1, D0, Topic=Client/data/test, Payload=...(15 bytes))
Client test sent DISCONNECT
```

**Reply by aze2201** [reply]
Actually this is working.
But I want to avoid to connect any  other topic which ```!=CN```. 
User can connect only topic name if equal to CN. Other should be reject.

**Reply by SergeTupchiy** [reply]
The example I shared will only allow to publish/subscribe if the first topic level equals `CN`. 
For example, it allows a user with `CN=foo` to pub/sub to topics `foo`, `foo/bar` but it will deny access for this user to topics like `bar/baz` etc.

To achieve the exact topic/CN match, you can remove topic wildcard (set `"%u"` instead of `"%u/#"`):
```
{allow, all, pubsub, ["%u"]}.
{deny, all}.
```
This will deny any publish/subscribe where CN != topic.

**Reply by aze2201** [reply]
Here is my rule in Gui > Modules > Internal ACL file.
```
{allow, all, pubsub, ["%u"]}.
{deny, all}.
```

Here is my cert
```openssl x509 -in /etc/mosquitto/certs/u1.crt  -text
...
Subject: CN = bridge1, ST = Some-State, C = de, O = contab, OU = fff
...
```

So, using this I have able to connect only bridge1 topic with this cert. 
But I can subscribe to bridge1111 topic as well which I want to avoid.

```
1683129147: Bridge local.xxx.com.m1m0 sending SUBSCRIBE (Mid: 2, Topic: bridge111, QoS: 1, Options: 0x00)
```

**Reply by aze2201** [reply]
```acl_nomatch = Deny```

**Comment by aze2201** [comment]
I see now it connects, but not subcribes.  Can I force to disconnect ? or this is out of scope of ACL.

```
acl_deny_action = disconnect
acl_nomatch = Deny
```

**Reply by SergeTupchiy** [reply]
acl_deny_action = disconnect should work, perhaps you just need to set it per a relevant zone, for example:
etc/zones.conf:
```
zone.external.acl_deny_action = disconnect
```
